src/System.Text.Primitives/System/Text/Formatting/Internal/FormattingHelpers.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text
+{
+    // All the helper methods in this class assume that the by-ref is valid and that there is
+    // enough space to fit the items that will be written into the underlying memory. The calling
+    // code must have already done all the necessary validation.
+    internal static class FormattingHelpers
+    {
+        private const int FractionDigits = 7;
+        private const string HexTable = "0123456789abcdef";
+
+        #region UTF-8 Helper methods
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe void WriteHexByte(byte value, ref byte buffer, int index)
+        {
+            Unsafe.Add(ref buffer, index) = (byte)HexTable[value >> 4];
+            Unsafe.Add(ref buffer, index + 1) = (byte)HexTable[value & 0xF];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int WriteFractionDigits(long value, int digitCount, ref byte buffer, int index)
+        {
+            for (var i = FractionDigits; i > digitCount; i--)
+                value = DivMod(value, 10, out long m);
```

Comment:
In most cases, that out param is used since that is the purpose of this function. However, this particular usage is actually leftover from when I was using a custom implementation of DivMod10 that was faster but obscure and only worked on positive integers up to 1 billion or so. I ended up reverting to the Math.DivRem implementation for easier maintenance.

I will switch this to a straight division since my optimized DivMod10 is gone now.